### ISSUE TITLE
Figure out whether Hugging Face transformers models can be restored from cache on CI runners

### ISSUE THREAD

**Issue by koernerfelicia** [RasaHQ/rasa#7620]
**Description of Problem**:
Downloading models from the Hugging Face transformers repository can take more than the 120s allotted for a pytest to run, resulting in a failed test due to timeout. This is especially problematic for CI, as the cached models are deleted after the runner is shut down.

We can download all the models in a fixture like [so](https://github.com/RasaHQ/rasa/pull/7365/commits/3bb1547acd6df8e659b3399f4b995234ca2e7fb7#diff-a5745664c84fd73bdc83ddbb9db8f145ec8c8a50ee2cef916f99125fbd084fdaR14-R24)

However, this also takes time (downloading the `bert` model takes  ~10 minutes when I run locally). For this reason, it may be better to restore the models from a cache.

See investigation of the issue [here](https://github.com/RasaHQ/rasa/pull/7365/).

**Overview of the Solution**:
Models should be restored from a cache. We already cache models after any download, and first check the cache before loading a model. Therefore, the rasa code and tests could remain the same, if the cached models are restored to the default cache location.

**Note**: The setup fixture should only be called once, to avoid multiple attempts to write the lockfile. This means it must be done before the test jobs are run in parallel.

The models to be downloaded can be found under: `rasa.nlu.utils.hugging_face.registry` and should include every `Tokenizer` and `Model` for every `model_name` and default `model_weights` for that `model_name`. More details can be found in the fixture linked above. Note however, that we may want to expand this list in the future for different tests.

**Definition of Done**:
- [ ] Tests can run on CI

**Comment by koernerfelicia** [comment]
[Example](https://github.com/RasaHQ/rasa/runs/1453373729?) of run where tests crashed. Note you can also reproduce the timeout locally by deleting the Hugging Face crash (something like `rm -rf ~/.cache/huggingface/transformers/*`).
I think the warning in the logs about 10% of memory being allocated can be ignored (source: Tanja) or at least isn't the reason why the runners are crashing. I did some local memory profiling [here](https://github.com/RasaHQ/rasa/pull/7365/#issuecomment-741092656) that supports the warning appearing.

**Comment by tczekajlo** [comment]
@koernerfelicia Do you know how much storage space takes files in the `~/.cache/huggingface/transformers/*`?

**Comment by koernerfelicia** [comment]
Double-checking atm, but I'm afraid it's something like ~6G. If that's too large I can split up into tests with large models to be run locally and tests with smaller models to be run on CI. Do you have a rough limit for the size?

**Comment by tczekajlo** [comment]
Thanks a lot for checking.

There are some issues with using the GH cache. 

- Cache limits
> A repository can have up to 5GB of caches. Once the 5GB limit is reached, older caches will be evicted based on when the cache was last accessed. Caches that are not accessed within the last week will also be evicted.

It means that other caches couldn't be used (the size of files in `~/.cache/huggingface/transformers/` is too big anyway)

- The tests have to pass without cache. The GH cache is uploaded in a post-task when the job finished with success thus the tests have to pass even if there is no cache for them

**Comment by koernerfelicia** [comment]
Thank you for looking into this! So if I am understanding correctly, cache won't work either way because of point two (so even reducing the total size of the files in `~/.cache/huggingface/transformers/` won't help).

In that case, how would you feel about the other option with a fixture that downloads the models before the tests happen? Is there a rough time limit you wouldn't want to exceed for tests? I think some of the models are quite small and fairly quick to download, so we could also limit CI tests to those models, and test the full suite of models only locally.

**Comment by tczekajlo** [comment]
> Thank you for looking into this! So if I am understanding correctly, cache won't work either way because of point two (so even reducing the total size of the files in `~/.cache/huggingface/transformers/` won't help).
> 

Yes, this time GH cache won't help.

> In that case, how would you feel about the other option with a fixture that downloads the models before the tests happen? Is there a rough time limit you wouldn't want to exceed for tests? I think some of the models are quite small and fairly quick to download, so we could also limit CI tests to those models, and test the full suite of models only locally.

Is there a way to download the models without using a fixture? e.g. execute some command? 
The other question, do you know how much time the tests last themselves without including the time that is needed to download the models?

**Comment by koernerfelicia** [comment]
> Is there a way to download the models without using a fixture?
Yes, we should be able to do with some lines of python, or a shell script, or whatever really. If we download the files, name them the right thing (some sort of hash of the model name IIRC), move them to the expected place, we're good.

Once the models are downloaded the tests are well under the timeout. I ran that whole module for `lm_featurizer` just now (35 tests total), and it took 2m2s

**Comment by tczekajlo** [comment]
If it's something easy to do I would prefer to have a separate script that we can use to download the models and execute the script before tests run.

Something like we do with MITIE:

e.g. https://github.com/RasaHQ/rasa/blob/957de844129011b9a346fb9d0154e1414dedba26/Makefile#L124

**Comment by koernerfelicia** [comment]
Okay! I'll write a script for that. Should I just create a PR with changes to the Makefile?

**Comment by tczekajlo** [comment]
> Okay! I'll write a script for that. Should I just create a PR with changes to the Makefile?

Yes, please create a new PR 😄  Thanks a lot 🙌🏻

**Comment by tczekajlo** [comment]
> Okay! I'll write a script for that. Should I just create a PR with changes to the Makefile?

@koernerfelicia 👋🏻  I'll close the issue, please mention me in a PR with the script if it's ready 😄
